test/Microsoft.AspNetCore.Mvc.Razor.Test/Compilation/ViewsFeatureProviderTest.cs

```diff
@@ -141,6 +141,74 @@ public void PopulateFeature_ReturnsViewsFromAllAvailableApplicationParts()
                 });
         }
 
+        [Fact]
+        public void PopulateFeature_PrefersViewsFromEntryPointAssembly()
+        {
+            // Arrange
+            var part1 = new AssemblyPart(typeof(ViewsFeatureProvider).Assembly, isEntryPoint: true);
+            var item1 = new TestRazorCompiledItem(typeof(StringBuilder), "mvc.1.0.view", "/Areas/Admin/Views/Shared/_Layout.cshtml", new object[] { });
+
+            var part2 = new AssemblyPart(GetType().Assembly);
+            var item2 = new TestRazorCompiledItem(typeof(string), "mvc.1.0.view", "/Areas/Admin/Views/Shared/_Layout.cshtml", new object[] { });
+            var item3 = new TestRazorCompiledItem(typeof(string), "mvc.1.0.view", "/Areas/Admin/Views/Shared/_Partial.cshtml", new object[] { });
+
+            var items = new Dictionary<AssemblyPart, IReadOnlyList<RazorCompiledItem>>
+            {
+                { part1, new[] { item1 } },
+                { part2, new[] { item2, item3, } },
+            };
+
+            var featureProvider = new TestableViewsFeatureProvider(items, attributes: new Dictionary<AssemblyPart, IEnumerable<RazorViewAttribute>>());
+            var partManager = new ApplicationPartManager();
+            partManager.ApplicationParts.Add(part1);
+            partManager.ApplicationParts.Add(part2);
+            partManager.FeatureProviders.Add(featureProvider);
+            var feature = new ViewsFeature();
+
+            // Act
+            partManager.PopulateFeature(feature);
+
+            // Assert
+            Assert.Collection(feature.ViewDescriptors.OrderBy(f => f.RelativePath, StringComparer.Ordinal),
+                view => Assert.Same(item1, view.Item),
+                view => Assert.Same(item3, view.Item));
+        }
+
+        [Fact]
+        public void PopulateFeature_ThrowsIfMultipleNonEntryPointViewsHaveTheSamePath()
+        {
+            // Arrange
+            var expected = string.Join(
+                Environment.NewLine,
+                @"More than one compiled view found that resolve to the same location:",
+                "Microsoft.AspNetCore.Mvc.Razor.Test: /Areas/Admin/Views/Shared/_layout.cshtml",
+                "Microsoft.AspNetCore.Mvc.Razor: /Areas/Admin/Views/Shared/_Layout.cshtml");
```

Comment:
Yeah I think this needs more care. 


We might want to gather **all** possible errors and printing them all together. We do this in places like routing because you don't want to play whackamole with errors.

```
The application's compiled views have conflicts. This means that multiple compiled view assemblies 
provide a view for the same path. This is also possible when an assembly has been renamed
and the application directory contains old publish output.

The following conflicts were detected for compiled views:
Path: '/Areas/Admin/Views/Shared/_layout.cshtml'
    Assembly: 'Foo'
    Assembly: 'Bar'

Path: '/Areas/Admin/Views/Shared/_Login.cshtml'
    Assembly: 'Foo'
    Assembly: 'Bar'
```

It might also be a good idea to terminate after about 10 or so errors because past that it gets really unwieldy.

You could also think about separating the discovery into a fast-path and a slow-path. If you run into an error on the fast path you go to the slow path with actually collects all of this information.